nixos/modules/services/web-apps/discourse.nix

```diff
@@ -0,0 +1,1079 @@
+{ config, options, lib, pkgs, utils, ... }:
+
+let
+  json = pkgs.formats.json {};
+
+  cfg = config.services.discourse;
+
+  postgresqlPackage = if config.services.postgresql.enable then
+                        config.services.postgresql.package
+                      else
+                        pkgs.postgresql;
+
+  # We only want to create a database if we're actually going to connect to it.
+  databaseActuallyCreateLocally = cfg.database.createLocally && cfg.database.host == null;
+
+  tlsEnabled = (cfg.enableACME
+                || cfg.sslCertificate != null
+                || cfg.sslCertificateKey != null);
+
+  discourseEnv = {
+    HOME = "/run/discourse/home";
+    RAILS_ENV = "production";
+    UNICORN_LISTENER = "/run/discourse/sockets/unicorn.sock";
+    UNICORN_TIMEOUT = builtins.toString cfg.unicornTimeout;
+    UNICORN_SIDEKIQS = builtins.toString cfg.sidekiqProcesses;
+  };
+
+  discourseDeps = [
+    postgresqlPackage
+
+    # For backups, themes and assets
+    cfg.package.ruby
+    pkgs.gzip
+    pkgs.gnutar
+    pkgs.git
+    pkgs.brotli
+
+    # Misc required system utils
+    pkgs.which
+    pkgs.procps       # For ps and kill
+    pkgs.util-linux   # For renice
+    pkgs.gawk
+
+    # Image optimization
+    pkgs.imagemagick
+    pkgs.optipng
+    pkgs.pngquant
+    pkgs.libjpeg
+    pkgs.jpegoptim
+    pkgs.gifsicle
+    pkgs.nodePackages.svgo
+  ];
+
+  discourse-rake = pkgs.runCommandNoCC "discourse-rake" {
+    nativeBuildInputs = [ pkgs.makeWrapper ];
+  } ''
+    mkdir -p $out/bin
+    makeWrapper ${cfg.package.rubyEnv}/bin/rake $out/bin/discourse-rake \
+        ${lib.concatStrings (lib.mapAttrsToList (name: value: "--set ${name} '${value}' ") discourseEnv)} \
+        --prefix PATH : ${lib.makeBinPath discourseDeps} \
```

Comment:
Is there a particular thing preventing us from wrapping the binary with these dependencies directly in the discourse package fixupPhase?

Reply:
It's not really dependent on any runtime configuration, not anymore (and rake shouldn't need access to the `UNICORN_`-variables), so I moved `discourse-rake` into the package instead. Note that it uses `rake` from the `rubyEnv`, though, so it's not possible to do in the `fixupPhase`.